paywall/src/data-iframe/blockchainHandler/purchaseKey.js

```diff
@@ -0,0 +1,168 @@
+import { getNetwork } from './network'
+import { getAccount } from './account'
+import ensureWalletReady from './ensureWalletReady'
+import { TRANSACTION_TYPES } from '../../constants'
+import pollForChanges from './pollForChanges'
+
+export async function purchaseKey(walletService, window, lock, amountToSend) {
+  await ensureWalletReady(window)
+  const account = getAccount()
+
+  return walletService.purchaseKey(lock, account, amountToSend)
+}
+
+/**
+ * Listens for the next time an event occurs, and resolves. If an error occurs first, it rejects
+ *
+ * @param {*} service either web3Service or walletService
+ * @param {string} event event to listen for
+ */
+function resolveOnEvent(service, event) {
+  let resolved = false
+  return new Promise((resolve, reject) => {
+    service.once('error', e => {
+      if (resolved) return
+      reject(e)
+    })
+    service.once(event, (...args) => resolve(args))
+  })
+}
+
+/**
+ * Does a basic shallow equal for transaction objects to see if any changes have occurred
+ * This is used to decide whether to trigger an update to the data
+ * @param {*} a old transaction object
+ * @param {*} b new transaction object
+ */
+function sortOfEqual(a, b) {
+  const aKeys = Object.keys(a).sort()
+  const bKeys = Object.keys(b).sort()
+  if (aKeys.length !== bKeys.length) return false
+  if (aKeys.filter((key, index) => bKeys[index] !== key).length) return false
+  if (aKeys.filter(key => a[key] !== b[key]).length) return false
+  return true
+}
+
+/**
+ * Listen for the transaction.pending and transaction.new events
+ *
+ * It calls `onTransactionUpdate` whenever one of these events is emitted
+ *
+ * It returns the updated list of all transactions, keys, and the current transaction status
+ */
+export async function processKeyPurchaseTransaction({
+  walletService,
+  lock,
+  existingTransactions,
+  existingKeys,
+  onTransactionUpdate,
+}) {
+  const account = getAccount()
+  const transactions = {
+    ...existingTransactions,
```

Comment:
Can we just name the props `transactions`?

Reply:
On line 107 we modify it, so I don't want to cause side effects that could be hard to debug later.